### PR TITLE
[docs] Add source control code references docs

### PR DESCRIPTION
**Pull request by benpankow**
## Summary

Expands the code references guide to include info on attaching references to git source control on cloud and OSS.

## Test Plan

Vercel.

### CODE REVIEW

**Comment by benpankow** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/22319?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#22319** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/22319?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#22318** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/22318?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#21755** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21755?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @benpankow and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: docs/content/guides/dagster/code-references.mdx [hunk 1]
```diff
@@ -113,3 +113,124 @@ Each of the code references to `manual_references.py` will be visible in the **A
...
+
+---
+
+## Converting code references to link to a git repository
+
+In a local context, it is useful to specify local code references in order to navigate directly to the source code of an asset. However, in a production environment, you may want to link to the source control repository where the code is stored.
+
+### In Dagster Plus
+
+If using Dagster Plus, you can use the `link_code_references_to_git_if_cloud` utility to conditionally convert local file code references to source control links. This utility will automatically detect if your code is running in a Dagster Cloud environment and convert local file code references to source control links, pointing at the commit hash of the code running in the current deployment.
```
Comment (sryza) [inline docs/content/guides/dagster/code-references.mdx:125]: Can we use `PyObject` to link to the API doc?

File: docs/content/guides/dagster/code-references.mdx [hunk 2]
```diff
@@ -113,3 +113,124 @@ Each of the code references to `manual_references.py` will be visible in the **A
...
+
+```python
```
Comment (sryza) [inline docs/content/guides/dagster/code-references.mdx:127]: This snippet should go in a .py file.

**Review by sryza — CHANGES_REQUESTED** [review]
Two comments - otherwise looks good to go!

File: docs/content/guides/dagster/code-references.mdx (hunk 2 above)
Comment (benpankow) [reply docs/content/guides/dagster/code-references.mdx:127]: Right now our docs build doesn't include cloud packages, so this won't work in a snippet file. Can file as a todo?

File: docs/content/guides/dagster/code-references.mdx (hunk 1 above)
Comment (benpankow) [reply docs/content/guides/dagster/code-references.mdx:125]: As below - cloud packages aren't worked into the docs build

File: docs/content/guides/dagster/code-references.mdx (hunk 2 above)
Comment (sryza) [reply docs/content/guides/dagster/code-references.mdx:127]: Ah yeah - worth adding a comment to explain why we're doing this.

File: docs/content/guides/dagster/code-references.mdx [hunk 3]
```diff
@@ -113,3 +113,124 @@ Each of the code references to `manual_references.py` will be visible in the **A
...
+            local_file_anchor=Path(__file__),
+            file_anchor_path_in_repository="src/repo.py",
+        ),
+    )
+)
+```
+
+### In any Dagster environment
+
+The <PyObject module="dagster" object="dagster._core.definitions.metadata.link_code_references_to_git"/> utility allows you to convert local file code references to source control links. You'll need to provide the base URL of your git repository, the branch or commit hash, and a path to the repository root locally.
```
Comment (sryza) [inline docs/content/guides/dagster/code-references.mdx:164]:
```suggestion
The <PyObject module="dagster" object="link_code_references_to_git"/> utility allows you to convert local file code references to source control links. You'll need to provide the base URL of your git repository, the branch or commit hash, and a path to the repository root locally.
```

File: docs/content/guides/dagster/code-references.mdx [hunk 4]
```diff
@@ -113,3 +113,124 @@ Each of the code references to `manual_references.py` will be visible in the **A
...
+```
+
+You may choose to conditionally apply this transformation based on the environment in which your Dagster code is running. For example, you could use an environment variable to determine whether to link to local files or to a source control repository:
+
+```python file=/guides/dagster/code_references/link_to_source_control_conditional.py
+import os
+from pathlib import Path
+
+from dagster import Definitions, asset
+from dagster._core.definitions.metadata import (
```
Comment (sryza) [inline docs/content/guides/dagster/code-references.mdx:206]:
```suggestion
from dagster import (
```

File: examples/docs_snippets/docs_snippets/guides/dagster/code_references/link_to_source_control_conditional.py [hunk 5]
```diff
@@ -0,0 +1,34 @@
+import os
+from pathlib import Path
+
+from dagster import Definitions, asset
+from dagster._core.definitions.metadata import (
```
Comment (sryza) [inline examples/docs_snippets/docs_snippets/guides/dagster/code_references/link_to_source_control_conditional.py:5]:
```suggestion
from dagster import (
```

File: docs/content/guides/dagster/code-references.mdx [hunk 6]
```diff
@@ -27,7 +27,7 @@ To complete the steps in this guide, you'll need:
 
 ---
 
-## Automatically attaching code references to asset definitions
+## Automatically attaching local file code references to asset definitions
 
 ### Assets defined in Python
```
Comment (sryza) [inline docs/content/guides/dagster/code-references.mdx:33]: I can't comment on the line below, but we should use `PyObject` and avoid an import that goes through a private path.

**Review by sryza — COMMENTED** [review]
Looks good, but heed final comments.

Also, do we have these in an rst file?

File: docs/content/guides/dagster/code-references.mdx [hunk 7]
```diff
@@ -27,11 +27,11 @@ To complete the steps in this guide, you'll need:
...
 
-To automatically attach code references to Python assets' function definitions, you can use the `dagster._core.definitions.metadata.with_source_code_references` utility. Any asset definitions passed to the utility will have their source file attached as metadata.
+To automatically attach code references to Python assets' function definitions, you can use the <PyObject module="dagster" object="with_source_code_references"/> utility. Any asset definitions passed to the utility will have their source file attached as metadata.
```
Comment (PedramNavid) [inline docs/content/guides/dagster/code-references.mdx:34]: This PyObject link doesn't work

File: docs/content/guides/dagster/code-references.mdx [hunk 8]
```diff
@@ -65,7 +65,7 @@ Dagster's dbt integration can automatically attach references to the SQL files b
```
Comment (PedramNavid) [inline docs/content/guides/dagster/code-references.mdx:65]:
```
A link to the asset's source in `with_source_code_references.py` will then be visible in the **Asset Catalog** view in the Dagster UI:
```

There's no reference to `with_source_code_references.py` in the above prose/code.

Comment (PedramNavid) [reply docs/content/guides/dagster/code-references.mdx:65]: Also that image is way too big

File: docs/content/guides/dagster/code-references.mdx [hunk 9]
```diff
@@ -113,3 +113,127 @@ Each of the code references to `manual_references.py` will be visible in the **A
...
+            local_file_anchor=Path(__file__),
+            file_anchor_path_in_repository="src/repo.py",
+        ),
+    )
+)
+```
+
+### In any Dagster environment
+
+The <PyObject module="dagster" object="link_code_references_to_git"/> utility allows you to convert local file code references to source control links. You'll need to provide the base URL of your git repository, the branch or commit hash, and a path to the repository root locally.
```
Comment (PedramNavid) [inline docs/content/guides/dagster/code-references.mdx:166]: This PyObject also doesn't work

**Review by PedramNavid — COMMENTED** [review]
Late to the game but these docs still need work before we can publish 1.8 I think? None of the API links work.

**Pull request merged**